fastai/basic_train.py

```diff
@@ -237,31 +238,34 @@ def export(self, fname:PathOrStr='export.pkl', destroy=False):
             xtra = dict(normalize=self.data.norm.keywords) if getattr(self.data, 'norm', False) else {}
             state['data'] = self.data.valid_ds.get_state(**xtra)
             state['cls'] = self.__class__
-            try_save(state, self.path, fname)
+            try_save(state, self.path, file)
         if destroy: self.destroy()
 
-    def save(self, name:PathOrStr, return_path:bool=False, with_opt:bool=True):
-        "Save model and optimizer state (if `with_opt`) with `name` to `self.model_dir`."
-        self._test_writeable_path()
+    def save(self, file:PathLikeOrBinaryStream=None, return_path:bool=False, with_opt:bool=True):
+        """Save model and optimizer state (if `with_opt`) with `file` to `self.model_dir`.
+        If `file` is a binary stream (buffer or file), write to it."""
+        if is_pathlike(file): self._test_writeable_path()
         if rank_distrib(): return # don't save if slave proc
-        path = self.path/self.model_dir/f'{name}.pth'
+        target = self.path/self.model_dir/f'{file}.pth' if is_pathlike(file) else file
```

Comment:
do we want to validate that file is actually a buffer if it's not str or pathlib?

Reply:
Well, Python is not strongly typed so `typing` doesn't really take care of it (IDEs will issue a warning).
I followed the `torch` convention here where they check if it's path-like and assume file-like otherwise.

Reply:
we could be more defensive by perhaps validating that `file` arg is one of the 4 types as the first thing in the function. But, perhaps, it's highly unlikely this will happen.

And I'm not sure @sgugger will agree as it'll add an extra line in several functions. So probably it's just fine for now.

Reply:
Honestly, I don't it's worth the trouble. We could be defensive all over the place, not just here. Recall the following line from the "how to contribute" guide:
> Contributing to Parkinson’s law of triviality has negative consequences for a project. Let’s focus on deep learning!

Reply:
I think the error message PyTorch will throw if `file` is of the wrong type so it's no necessary to worry here.